core/src/main/scala/kafka/controller/PartitionStateMachine.scala

```diff
@@ -146,27 +146,46 @@ class PartitionStateMachine(config: KafkaConfig,
    */
   private def doHandleStateChanges(partitions: Seq[TopicAndPartition], targetState: PartitionState,
                            partitionLeaderElectionStrategyOpt: Option[PartitionLeaderElectionStrategy] = None): Unit = {
+    val stateChangeLog = stateChangeLogger.withControllerEpoch(controllerContext.epoch)
     partitions.foreach(partition => partitionState.getOrElseUpdate(partition, NonExistentPartition))
     val (validPartitions, invalidPartitions) = partitions.partition(partition => isValidTransition(partition, targetState))
     invalidPartitions.foreach(partition => logInvalidTransition(partition, targetState))
     targetState match {
       case NewPartition =>
-        validPartitions.foreach(partition => partitionState.put(partition, NewPartition))
+        validPartitions.foreach { partition =>
+          stateChangeLog.trace(s"Changed partition $partition state from ${partitionState(partition)} to $targetState with " +
+            s"assigned replicas ${controllerContext.partitionReplicaAssignment(partition).mkString(",")}")
+          partitionState.put(partition, NewPartition)
```

Comment:
Hmm not sure if this would actually work. Some of these concepts don't even exist in certain states. 
- NewPartition has no leader or isr.
- NonExistentPartition has no replicas, leader, or isr.

Reply:
Ok. Maybe in the case where we transition to OnlinePartition, we can just log the whole leaderAndIsr instead of just the leader.

Reply:
Done